mempool/priority/mempool.go

```diff
@@ -0,0 +1,819 @@
+	for i, tx := range blockTxs {
+		// Add successful committed transactions to the cache (if they are not
+		// already present).  Transactions that failed to commit are removed from
+		// the cache unless the operator has explicitly requested we keep them.
+		if deliverTxResponses[i].Code == abci.CodeTypeOK {
+			_ = txmp.cache.Push(tx)
+		} else if !txmp.config.KeepInvalidTxsInCache {
+			txmp.cache.Remove(tx)
+		}
+
+		// Regardless of success, remove the transaction from the mempool.
+		_ = txmp.removeTxByKey(tx.Key())
+	}
+
+	txmp.purgeExpiredTxs(blockHeight)
+
+	// If there any uncommitted transactions left in the mempool, we either
+	// initiate re-CheckTx per remaining transaction or notify that remaining
+	// transactions are left.
+	size := txmp.Size()
+	txmp.metrics.Size.Set(float64(size))
+	txmp.metrics.SizeBytes.Set(float64(txmp.SizeBytes()))
+	if size > 0 {
+		if txmp.config.Recheck {
+			txmp.recheckTransactions()
+		} else {
+			txmp.notifyTxsAvailable()
+		}
+	}
+	return nil
+}
+
+// addNewTransaction handles the ABCI CheckTx response for the first time a
+// transaction is added to the mempool.  A recheck after a block is committed
+// goes to handleRecheckResult.
+//
+// If either the application rejected the transaction or a post-check hook is
+// defined and rejects the transaction, it is discarded.
+//
+// Otherwise, if the mempool is full, check for lower-priority transactions
+// that can be evicted to make room for the new one. If no such transactions
+// exist, this transaction is logged and dropped; otherwise the selected
+// transactions are evicted.
+//
+// Finally, the new transaction is added and size stats updated.
+func (txmp *TxMempool) addNewTransaction(wtx *WrappedTx, checkTxRes *abci.ResponseCheckTx) {
+	var err error
+	if txmp.postCheckFn != nil {
+		err = txmp.postCheckFn(wtx.tx, checkTxRes)
+	}
+
+	if err != nil || checkTxRes.Code != abci.CodeTypeOK {
+		txmp.logger.Debug(
+			"rejected bad transaction",
+			"priority", wtx.Priority(),
+			"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+			"peer_id", wtx.peers,
+			"code", checkTxRes.Code,
+			"post_check_err", err,
+		)
+
+		txmp.metrics.FailedTxs.Add(1)
+
+		// Remove the invalid transaction from the cache, unless the operator has
+		// instructed us to keep invalid transactions.
+		if !txmp.config.KeepInvalidTxsInCache {
+			txmp.cache.Remove(wtx.tx)
+		}
+
+		return
+	}
+
+	priority := getTxPriority(int64(checkTxRes.GasFees), checkTxRes.GasWanted)
+	sender := checkTxRes.Address
+
+	// Disallow multiple concurrent transactions from the same sender assigned
+	// by the ABCI application. As a special case, an empty sender is not
+	// restricted.
+	if len(sender) > 0 {
+		elt, ok := txmp.txBySender[string(sender)]
+		if ok {
+			w := elt.Value.(*WrappedTx)
+			txmp.logger.Debug(
+				"rejected valid incoming transaction; tx already exists for sender",
+				"tx", fmt.Sprintf("%X", w.tx.Hash()),
+				"sender", sender,
+			)
+			return
```

Comment:
[not blocking]
previously we were adding more specific errors to the response here, can we not do that anymore or is it just not needed?

Reply:
the mempool error was removed from checktx response. it seems like it was a hack that was removed as part of 38. I can add it to the debug but not sure how to best propagate it. do you have thoughts on this?

Reply:
we likely will need to make sure we're returning good errors at some point, but I'll open an issue and we can handle later

Reply:
In practice, the cosmos sdk never uses the sender field right @tac0turtle ?

Reply:
correct as a transaction could have many signers